src/libraries/System.Private.CoreLib/src/System/BitConverter.cs

```diff
@@ -741,9 +741,9 @@ public static bool ToBoolean(byte[] value, int startIndex)
             if (value == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.value);
             if (startIndex < 0)
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startIndex, ExceptionResource.ArgumentOutOfRange_Index);
-            if (startIndex > value.Length - 1)
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startIndex, ExceptionResource.ArgumentOutOfRange_Index); // differs from other overloads, which throw base ArgumentException
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startIndex, ExceptionResource.ArgumentOutOfRange_IndexMustBeLess);
+            if (startIndex >= value.Length)
```

Comment:
Hi 👋

Probably, https://github.com/dotnet/perf-autofiling-issues/issues/4738 and https://github.com/dotnet/runtime/issues/68338 are my bad. Here I've changed to `>=` so that it'd be more consistent(`>=`, `...IndexMustBeLess`) and more readable. 

But after seeing this benchmark and being able to reproduce it, I've reverted my commit and compared it's performance to upstream. I've found out that it's my commit that causing this and somehow, on jit asm level `>=` approx two times slower then `> len - 1`. So I've fixed it and got the benchmark result:

|                                    Method |        Job |                                                                                                  Toolchain |       Mean |     Error |     StdDev |     Median |        Min |        Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------------------------------------ |----------- |----------------------------------------------------------------------------------------------------------- |-----------:|----------:|-----------:|-----------:|-----------:|-----------:|------:|--------:|----------:|------------:|
|                    ReadStructAndReverseBE | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  |  11.345 ns | 0.1365 ns |  0.1277 ns |  11.394 ns |  11.116 ns |  11.477 ns |  0.98 |    0.01 |         - |          NA |
|                    ReadStructAndReverseBE | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  11.566 ns | 0.0608 ns |  0.0539 ns |  11.568 ns |  11.472 ns |  11.660 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
|                    ReadStructAndReverseLE | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  |   1.992 ns | 0.0726 ns |  0.0679 ns |   1.981 ns |   1.831 ns |   2.133 ns |  1.01 |    0.03 |         - |          NA |
|                    ReadStructAndReverseLE | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   1.983 ns | 0.0088 ns |  0.0078 ns |   1.984 ns |   1.962 ns |   1.994 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
|                  ReadStructFieldByFieldBE | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  |  10.517 ns | 0.5510 ns |  0.5896 ns |  10.242 ns |   9.892 ns |  12.019 ns |  1.00 |    0.07 |         - |          NA |
|                  ReadStructFieldByFieldBE | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  10.478 ns | 0.4902 ns |  0.5645 ns |  10.132 ns |   9.623 ns |  11.228 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
|                  ReadStructFieldByFieldLE | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  |   9.229 ns | 0.0381 ns |  0.0337 ns |   9.227 ns |   9.180 ns |   9.296 ns |  1.01 |    0.00 |         - |          NA |
|                  ReadStructFieldByFieldLE | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   9.139 ns | 0.0066 ns |  0.0055 ns |   9.136 ns |   9.133 ns |   9.153 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
| ReadStructFieldByFieldUsingBitConverterLE | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  |  19.257 ns | 0.1194 ns |  0.0997 ns |  19.255 ns |  19.008 ns |  19.404 ns |  0.90 |    0.01 |         - |          NA |
| ReadStructFieldByFieldUsingBitConverterLE | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  21.446 ns | 0.1205 ns |  0.1127 ns |  21.434 ns |  21.258 ns |  21.603 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
| ReadStructFieldByFieldUsingBitConverterBE | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  |  30.687 ns | 0.0825 ns |  0.0732 ns |  30.695 ns |  30.544 ns |  30.804 ns |  0.98 |    0.00 |         - |          NA |
| ReadStructFieldByFieldUsingBitConverterBE | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  31.372 ns | 0.1749 ns |  0.1636 ns |  31.297 ns |  31.221 ns |  31.792 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
|                  MeasureReverseEndianness | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  | 402.172 ns | 0.5438 ns |  0.5086 ns | 402.114 ns | 401.325 ns | 402.919 ns |  0.91 |    0.00 |         - |          NA |
|                  MeasureReverseEndianness | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 441.049 ns | 2.4377 ns |  2.2803 ns | 440.656 ns | 438.690 ns | 445.914 ns |  1.00 |    0.00 |         - |          NA |
|                                           |            |                                                                                                            |            |           |            |            |            |            |       |         |           |             |
|                   MeasureReverseUsingNtoH | Job-FQAXKT | /runtime.fixed/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun  | 421.494 ns | 1.1445 ns |  0.9557 ns | 421.163 ns | 420.339 ns | 423.134 ns |  1.01 |    0.03 |         - |          NA |
|                   MeasureReverseUsingNtoH | Job-KKCPCD |        /runtime/artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 423.682 ns | 9.7031 ns | 11.1741 ns | 428.362 ns | 407.732 ns | 437.481 ns |  1.00 |    0.00 |         - |          NA |

Also see [here](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQ6BuYBOACMPAXjwHEBTYACgEoBuLXPAI2LMtoYzTiQE4qhAHwlm9PAHoJeAMYALMADsA5uTwBLYHmAB7PAANho/dr0RyagCbqAZjfTpeANg2KtFajXQBvdHn/aOACektIAVgCuEFoQesAKWpoA5BB4ijpaOgAOwOpQ6gBe5JYEAO5gQX4BvmgBdXhwAOxp5KV4AEpKljpQtAB0AHLkAB7USAA0eEgADPRV/gC+87JgwPJ4yzX1AU14AKx0G7UBS2gLQA===) the difference between `a >= b` and `a > b - 1` in jit asm.

Mainly this diff:

```
// a >= b case
    L0012: cmp esi, eax
    L0014: setge cl

// a > b - 1 case
    L0012: dec eax
    L0013: cmp esi, eax
    L0015: setg cl
```

Now the questions are

1. Should I open another PR with quickfix, or how is it supposed to be done? 
2. Maybe I should investigate why it's faster, is it so on most platforms? And try to put that as optimization into jit. Or is it too minor/unreliable/inconsistent throughout platforms/etc?

Reply:
> * Should I open another PR with quickfix, or how is it supposed to be done?

Yes, open a new PR.

> * Maybe I should investigate why it's faster, is it so on most platforms? And try to put that as optimization into jit. Or is it too minor/unreliable/inconsistent throughout platforms/etc?

You're certainly welcome to investigate.

It can be tricky to root cause these. I would hazard a guess the perf issue is not directly related to the change above but some other microarchitectural issue (say branch alignment). Though given that we're also seeing impact on arm64 perhaps it is something directly related after all.